### PR TITLE
*: move significant send to raft router

### PR DESCRIPTION
**Pull request by siddontang**
Now we use raft router and a significant sender to send messages to the raft store, this seems a little duplicated, so here move the significant sender to the raft router, so we can only use the raft router to send messages.

### CODE REVIEW

**Comment by siddontang** [comment]
PTAL @BusyJay @overvenus

File: src/raftstore/store/store.rs [hunk 1]
```diff
@@ -445,20 +445,23 @@ impl<T, C> Store<T, C> {
...
                     self.report_snapshot_status(region_id, to_peer_id, status);
                 }
-                Ok(SignificantMsg::Unreachable {
+                Ok(Msg::Unreachable {
                     region_id,
                     to_peer_id,
                 }) => if let Some(peer) = self.region_peers.get_mut(&region_id) {
                     peer.raft_group.report_unreachable(to_peer_id);
                 },
+                Ok(msg) => {
```
Comment (BusyJay) [inline src/raftstore/store/store.rs]: I prefer separating `SignificantMsg` from `Msg` giving that they doesn't share sender/receiver.

File: src/server/server.rs [hunk 2]
```diff
@@ -282,13 +287,15 @@ mod tests {
...
-            }
-        );
+        if let StoreMsg::Unreachable {
+            region_id: 0,
+            to_peer_id: 0,
+        } = significant_msg_receiver.try_recv().unwrap()
+        {
+            // nothing to do
+        } else {
+            assert!(false);
```
Comment (BusyJay) [inline src/server/server.rs]: It's meaningless, why not use `assert_eq` or just panic with meaningful message?

File: src/server/transport.rs [hunk 3]
```diff
@@ -55,16 +55,50 @@ pub trait RaftStoreRouter: Send + Clone {
...
+        })
+    }
 }
 
 #[derive(Clone)]
 pub struct ServerRaftStoreRouter {
     pub ch: SendCh<StoreMsg>,
+    // Some tests need Sync for ServerRaftStoreRouter, so here we use Arc + Mutex here.
+    // TODO: refactor the test and remove the limitation later.
+    pub significant_msg_sender: Arc<Mutex<Sender<StoreMsg>>>,
```
Comment (overvenus) [inline src/server/transport.rs]: How about `SyncSender`?

File: tests/raftstore/transport_simulate.rs [hunk 4]
```diff
@@ -29,6 +29,9 @@ use std::sync::atomic::*;
 
 pub trait Channel<M>: Send + Clone {
     fn send(&self, m: M) -> Result<()>;
+    fn significant_send(&self, _: M) -> Result<()> {
+        unreachable!()
```
Comment (overvenus) [inline tests/raftstore/transport_simulate.rs]: `unimplemented!()`

File: src/server/server.rs [hunk 5]
```diff
@@ -282,13 +287,15 @@ mod tests {
...
-            significant_msg_receiver.try_recv().unwrap(),
-            SignificantMsg::Unreachable {
-                region_id: 0,
-                to_peer_id: 0,
-            }
-        );
+        if let StoreMsg::Unreachable {
+            region_id: 0,
+            to_peer_id: 0,
+        } = significant_msg_receiver.try_recv().unwrap()
```
Comment (overvenus) [inline src/server/server.rs]: How about `assert_eq!()` ?

File: src/server/transport.rs (hunk 3 above)
Comment (siddontang) [reply src/server/transport.rs]: SyncSender may send failed with Full error, but here we must always send successfully unless shutdown.

**Comment by siddontang** [comment]
PTAL @BusyJay

File: src/server/transport.rs [hunk 6]
```diff
@@ -55,16 +55,50 @@ pub trait RaftStoreRouter: Send + Clone {
...
+            status: status,
+        })
+    }
 }
 
 #[derive(Clone)]
 pub struct ServerRaftStoreRouter {
     pub ch: SendCh<StoreMsg>,
+    // Some tests need Sync for ServerRaftStoreRouter, so here we use Arc + Mutex here.
+    // TODO: refactor the test and remove the limitation later.
```
Comment (BusyJay) [inline src/server/transport.rs]: It may be more reasonable to move the lock to tests.

**Comment by siddontang** [comment]
PTAL @BusyJay

**Comment by siddontang** [comment]
PTAL @overvenus

**Pull request merged**